components/List/src/MDCSelfSizingStereoCell.m

```diff
@@ -0,0 +1,464 @@
+ limitations under the License.
+ */
+
+#import "MDCSelfSizingStereoCell.h"
+
+#import <MDFInternationalization/MDFInternationalization.h>
+#import "MaterialInk.h"
+#import "MaterialTypography.h"
+
+static const CGFloat kVerticalMarginMin = 8.0;
+static const CGFloat kVerticalMarginMax = 16.0;
+static const CGFloat kHorizontalMargin = 16.0;
+
+static const CGFloat kImageSideLengthMedium = 40.0;
+static const CGFloat kImageSideLengthMax = 56.0;
+
+static const CGFloat kTitleColorOpacity = 0.87f;
+static const CGFloat kDetailColorOpacity = 0.6f;
+
+@interface MDCSelfSizingStereoCellLayout : NSObject
+
+@property (nonatomic, assign, readonly) CGFloat cellWidth;
+@property (nonatomic, assign, readonly) CGFloat calculatedHeight;
+@property (nonatomic, assign, readonly) CGRect textContainerFrame;
+@property (nonatomic, assign, readonly) CGRect titleLabelFrame;
+@property (nonatomic, assign, readonly) CGRect detailLabelFrame;
+@property (nonatomic, assign, readonly) CGRect leadingImageViewFrame;
+@property (nonatomic, assign, readonly) CGRect trailingImageViewFrame;
+
+- (instancetype)initWithLeadingImageView:(UIImageView *)leadingImageView
+                       trailingImageView:(UIImageView *)trailingImageView
+                           textContainer:(UIView *)textContainer
+                              titleLabel:(UILabel *)titleLabel
+                             detailLabel:(UILabel *)detailLabel
+                               cellWidth:(CGFloat)cellWidth;
+
+@end
+
+@interface MDCSelfSizingStereoCellLayout ()
+
+@property (nonatomic, assign) CGFloat cellWidth;
+@property (nonatomic, assign) CGFloat calculatedHeight;
+@property (nonatomic, assign) CGRect textContainerFrame;
+@property (nonatomic, assign) CGRect titleLabelFrame;
+@property (nonatomic, assign) CGRect detailLabelFrame;
+@property (nonatomic, assign) CGRect leadingImageViewFrame;
+@property (nonatomic, assign) CGRect trailingImageViewFrame;
+
+@end
+
+@implementation MDCSelfSizingStereoCellLayout
+
+- (instancetype)initWithLeadingImageView:(UIImageView *)leadingImageView
+                       trailingImageView:(UIImageView *)trailingImageView
+                           textContainer:(UIView *)textContainer
+                              titleLabel:(UILabel *)titleLabel
+                             detailLabel:(UILabel *)detailLabel
+                               cellWidth:(CGFloat)cellWidth {
+  self = [super init];
+  if (self) {
+    self.cellWidth = cellWidth;
+    [self assignFrameForLeadingImageView:leadingImageView];
+    [self assignFrameForTrailingImageView:trailingImageView];
+    [self assignFramesForTextContainer:textContainer
+                            titleLabel:titleLabel
+                           detailLabel:detailLabel];
+    self.calculatedHeight = [self calculateHeight];
+  }
+  return self;
+}
+
+- (void)assignFrameForLeadingImageView:(UIImageView *)leadingImageView {
+  CGSize size = [self sizeForImage:leadingImageView.image];
```

Comment:
do we handle the case for when the user uses RTL?

Reply:
All these methods for assigning the frames assume LTR. The cell's `layoutSubviews` implementation basically just flips the frames if it's RTL. Do you think this is sufficient? I actually wasn't sure. Seems to work, but there might be some gotchas I'm missing!

Reply:
Done in 64652f0f58c0521445e92d50910e32fb9bdee360

Reply:
hey if it works it works :)